### PR TITLE
Working Maestro Thruster test

### PR DESCRIPTION
**Pull request by irwineffect**

### CODE REVIEW

**Comment by irwineffect** [comment]
The test should pass once #104 is merged in.

File: src/movement/test/CMakeLists.txt [hunk 1]
```diff
@@ -1,5 +1,11 @@
-add_rostest_gtest(test_sabertooth_thruster
+add_rostest_gtest(test_maestro_thruster
+  maestro_launch.test
+  test_maestro_thruster.cpp)
+add_dependencies(test_maestro_thruster thruster_maestro)
+target_link_libraries(test_maestro_thruster serial test_tools ${catkin_LIBRARIES} ${GTEST_LIBRARIES})
```
Comment (bkallaher) [inline src/movement/test/CMakeLists.txt]: The naming for this executable don't match those of the sabertooth.

File: src/movement/test/test_maestro_thruster.cpp [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+    //open our serial port
+    mSerial.Open(testing_port.c_str(), B9600);
+
+    //initialize our publisher used for sending thruster messages
+    pub = n.advertise<robosub::thruster>("thruster", 1);
+
+    ROS_INFO("all setup, press enter");
+    //ros::Duration(4).sleep();
+    /////getchar();
```
Comment (bkallaher) [inline src/movement/test/test_maestro_thruster.cpp]: Do these comments need to be here?

File: src/movement/test/test_maestro_thruster.cpp [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+    //this variable so it knows what serial port to open
+    ros::param::set("ports/thruster", UUT_port);
+
+    //open our serial port
+    mSerial.Open(testing_port.c_str(), B9600);
+
+    //initialize our publisher used for sending thruster messages
+    pub = n.advertise<robosub::thruster>("thruster", 1);
+
+    ROS_INFO("all setup, press enter");
```
Comment (bkallaher) [inline src/movement/test/test_maestro_thruster.cpp]: There is no getch so enter isn't needed.

File: src/movement/test/test_maestro_thruster.cpp [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+#include "ros/ros.h"
+#include "robosub/thruster.h"
+#include "utility/serial.hpp"
+#include "utility/test_tools.hpp"
+
+rs::Serial mSerial;
+ros::Publisher pub;
+double byte_check(uint8_t byte2, uint8_t byte3);
+std::vector<uint8_t> channels;
+double max_speed = 0.6;
```
Comment (bkallaher) [inline src/movement/test/test_maestro_thruster.cpp]: Is there really a need for globals?

File: src/movement/test/test_maestro_thruster.cpp [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+
+    ROS_INFO("published message, hit enter to read serial");
+    ros::Duration(1).sleep();
+
+    //after we published the thruster message, the thruster module should have
+    //received the message, and then sent some data down the serial port, here
+    //we read that data and verify it is correct
+    for(unsigned int i = 0; i < maestro_msg.data.size(); ++i)
+    {
+
```
Comment (bkallaher) [inline src/movement/test/test_maestro_thruster.cpp]: Extra whitespace

**Review by bkallaher — CHANGES_REQUESTED** [review]
Fails in the linter...

File: src/movement/test/test_maestro_thruster.cpp (hunk 4 above)
Comment (skallaher) [reply src/movement/test/test_maestro_thruster.cpp]: At least the `max_speed` value was just temporary to get the test in a working state.

File: src/movement/test/CMakeLists.txt (hunk 1 above)
Comment (irwineffect) [reply src/movement/test/CMakeLists.txt]: fixed!

File: src/movement/test/test_maestro_thruster.cpp (hunk 4 above)
Comment (irwineffect) [reply src/movement/test/test_maestro_thruster.cpp]: Yeah, in the future there will be more tests, they will use these variables.

File: src/movement/test/test_maestro_thruster.cpp (hunk 5 above)
Comment (irwineffect) [reply src/movement/test/test_maestro_thruster.cpp]: fixed

File: src/movement/test/test_maestro_thruster.cpp (hunk 3 above)
Comment (irwineffect) [reply src/movement/test/test_maestro_thruster.cpp]: fixed

File: src/movement/test/test_maestro_thruster.cpp (hunk 2 above)
Comment (irwineffect) [reply src/movement/test/test_maestro_thruster.cpp]: fixed

File: src/movement/test/test_thruster_maestro.cpp [hunk 6]
```diff
@@ -0,0 +1,146 @@
+//This file is designed to test the serial_subscriber.cpp example, which
+//receives messages of the robosub/thruster type, and then sends the exact
+//received value through a serial port. This example sends a single thruster
+//message with several thrusters inside, and then listens on a serial port to
+//confirm that the received values are the same as the sent values.
```
Comment (skallaher) [inline src/movement/test/test_thruster_maestro.cpp]: This comment claims to be testing the `serial_subscriber` example.

File: src/movement/test/test_thruster_maestro.cpp [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+#include "utility/serial.hpp"
+#include "utility/test_tools.hpp"
+#include <vector>
+#include <string>
+
+rs::Serial mSerial;
+ros::Publisher pub;
+double byte_check(uint8_t byte2, uint8_t byte3);
+std::vector<uint8_t> channels;
+double max_speed = 0.6;
```
Comment (skallaher) [inline src/movement/test/test_thruster_maestro.cpp]: Will this be made into a parameter in future? As well as in the unit under test?

File: src/movement/test/test_thruster_maestro.cpp [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+
+//insert tests here
```
Comment (skallaher) [inline src/movement/test/test_thruster_maestro.cpp]: Perhaps describe the test being run here.

File: src/movement/test/test_thruster_maestro.cpp [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+TEST(Maestro, basicTest)
+{
+    //allocate an array of bytes that will be used to receive serial data from
+    //the serial port. I use the arbitrarily large size of
+    uint8_t maestro_data[256] = {0};
```
Comment (skallaher) [inline src/movement/test/test_thruster_maestro.cpp:28]: I had the same question in the sabertooth test. Why is this an arbitrary number? Isn't this a guaranteed size given correct responses?

File: src/movement/test/test_thruster_maestro.cpp [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+    //received the message, and then sent some data down the serial port, here
+    //we read that data and verify it is correct
+    for(unsigned int i = 0; i < maestro_msg.data.size(); ++i)
+    {
+        //read one byte from the serial port
+        int bytes_rxd = mSerial.Read(maestro_data, 4);
+        ROS_INFO("recieved %d bytes", bytes_rxd);
+        //mSerial.Read(maestro_data, 4);
+
+        ROS_WARN("raw_vals: %x, %x", maestro_data[2], maestro_data[3]);
```
Comment (skallaher) [inline src/movement/test/test_thruster_maestro.cpp]: Are these two outputs the correct logger level for what they are?

File: src/movement/test/test_thruster_maestro.cpp [hunk 11]
```diff
@@ -0,0 +1,146 @@
...
+        //compare and make sure the value recieved is equal to the value sent
+        double expected_speed = maestro_msg.data[i];
+        if(fabs(expected_speed) > max_speed)
+        {
+            expected_speed = max_speed * ((expected_speed < 0)? -1 : 1);
+        }
+        EXPECT_FLOAT_EQ(actual_speed, expected_speed);
+    }
+
+    ROS_INFO("test over, waiting");
```
Comment (skallaher) [inline src/movement/test/test_thruster_maestro.cpp]: What is this waiting on?

File: src/movement/test/test_thruster_maestro.cpp [hunk 12]
```diff
@@ -0,0 +1,146 @@
...
+}
+
+
+int main(int argc, char *argv[])
+{
+    //initialize the google testing framework
+    testing::InitGoogleTest(&argc, argv);
+
+    //initialize our ros node
+    ros::init(argc, argv, "maestro_launch_test");
```
Comment (skallaher) [inline src/movement/test/test_thruster_maestro.cpp]: Would this be good to rename to reflect the refactoring of the executable?

File: src/movement/test/test_thruster_maestro.cpp [hunk 13]
```diff
@@ -0,0 +1,146 @@
...
+    //fill them out
+    std::string UUT_port;
+    std::string testing_port;
+
+    //try to start the serial testbench. If successful, it will return 0, on
+    //failure it will return 1
+    if(serial_tb.Start(testing_port, UUT_port) != 0)
+    {
+        ROS_FATAL("could not start serial testbench, failing test");
+        exit(1);
```
Comment (skallaher) [inline src/movement/test/test_thruster_maestro.cpp:128]: Does exiting this way actually fail the test?

File: src/movement/test/thruster_maestro.test [hunk 14]
```diff
@@ -0,0 +1,6 @@
+<launch>
+    <test test-name="test_maestro_thruster" pkg="robosub" type="test_maestro_thruster" />
```
Comment (skallaher) [inline src/movement/test/thruster_maestro.test]: The type doesn't match the new executable name, is this intended?

**Review by skallaher — CHANGES_REQUESTED** [review]

File: src/movement/test/thruster_maestro.test (hunk 14 above)
Comment (bkallaher) [reply src/movement/test/thruster_maestro.test]: Fixed

**Review by bkallaher — APPROVED** [review]

File: src/movement/test/test_thruster_maestro.cpp (hunk 7 above)
Comment (bkallaher) [reply src/movement/test/test_thruster_maestro.cpp]: It should be in the control parameters

File: src/movement/test/test_thruster_maestro.cpp (hunk 13 above)
Comment (bkallaher) [reply src/movement/test/test_thruster_maestro.cpp:128]: Any exit value other than 0 will fail the test.

File: src/movement/test/test_thruster_maestro.cpp (hunk 9 above)
Comment (bkallaher) [reply src/movement/test/test_thruster_maestro.cpp:28]: When it comes to any type of read or write on a communication bus you should always allocate more than the expected data. In a single read the program could take in more than one packet. In many cases (IP for example) it is required that an application read in all data available.

File: src/movement/test/test_thruster_maestro.cpp (hunk 8 above)
Comment (bkallaher) [reply src/movement/test/test_thruster_maestro.cpp]: Added.

File: src/movement/test/test_thruster_maestro.cpp (hunk 11 above)
Comment (bkallaher) [reply src/movement/test/test_thruster_maestro.cpp]: @irwineffect: is it still in need of the wait at the end/does it even exist?

File: src/movement/test/test_thruster_maestro.cpp (hunk 10 above)
Comment (bkallaher) [reply src/movement/test/test_thruster_maestro.cpp]: Logging isn't done when the test is actually run so it really doesn't matter.

File: src/movement/test/test_thruster_maestro.cpp (hunk 12 above)
Comment (bkallaher) [reply src/movement/test/test_thruster_maestro.cpp]: As long as the name is unique there really isn't a need to change the running node name here.

**Review by skallaher — APPROVED** [review]

**Pull request merged**